Try in-cluster config before using localhost:8080

dims opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

When starting an e2e test in a pod in a cluster, if the host is
not specified in the command line, we default to using
'http://127.0.0.1:8080' currently. We should be discovering the
host/port using the in-cluster config and using that if
possible.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Fixes #53894

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

dims committed:
Try in-cluster config before using localhost:8080

When starting an e2e test in a pod in a cluster, if the host is
not specified in the command line, we default to using
'http://127.0.0.1:8080' currently. We should try the in-cluster
config, save it to a temporary file and use that with kubectl

dims commented:
/sig testing

gekart commented:
Specifying the InClusterConfig host only will not suffice to authorize kubectl (https://github.com/kubernetes/kubernetes/issues/53894#issuecomment-336615448)

dims commented:
@gekart : from what i am reading there is a valid kubeconfig at this point which kubectl is picking up but there's code that over rides the host, so am trying to do the least invasive thing and just making sure the host is set.

This is why i was asking which tests are failing.

Thanks,
Dims

timothysc commented:
There is already code in the init portion that does this.  In order for the kubectl tests to work there needs to be a the following:

Write the in-cluster kubeconfig to disk in default path so shell'd execution of kubectl picks it up.

dims commented:
/test pull-kubernetes-unit

dims commented:
/hold

dims commented:
/hold cancel

dims commented:
/test pull-kubernetes-e2e-kubeadm-gce

cblecker commented:
/test pull-kubernetes-bazel-build

dims commented:
@gekart @cblecker @timothysc - So i edited the Dockerfile in the repo that Grischa pointed out to include kubectl - https://github.com/dims/k8s-e2e-containerized/commit/81b1e2aa597388bd1bad6901931dc63904319168 and got the test working see example run of one test below (Same test was failing before my patch - http://paste.openstack.org/show/623726/)

```
[dims@bigbox 09:39] ~/go/src/github.com/ozdanborne/k8s-e2e-containerized ⟩ docker rmi ozdanborne/k8s-e2e:latest
Untagged: ozdanborne/k8s-e2e:latest
Deleted: sha256:c3f8239b5e0b8eea082f2e3aa144cbe23f085f6b5ca31129f5d0df8926755644
Deleted: sha256:b5499f4c5ea320e79a77578d916efe341f081dde1af9f8a660840efb03302363
Deleted: sha256:2d455ea642029871733d121e6b9a0e8497dac109959951f6fdca490e845ebaab
[dims@bigbox 09:40] ~/go/src/github.com/ozdanborne/k8s-e2e-containerized ⟩ ls bin
ls: cannot access 'bin': No such file or directory
[dims@bigbox 09:40] ~/go/src/github.com/ozdanborne/k8s-e2e-containerized ⟩ make
mkdir -p bin
cp /home/dims/go/src/k8s.io/kubernetes/_output/local/bin/linux/amd64/e2e.test ./bin/
mkdir -p bin
cp /home/dims/go/src/k8s.io/kubernetes/_output/local/bin/linux/amd64/kubectl ./bin/
docker build -t ozdanborne/k8s-e2e .
Sending build context to Docker daemon  207.9MB
Step 1/8 : FROM golang
 ---> 99e596fc807e
Step 2/8 : LABEL maintainer "dan@projectcalico.org"
 ---> Using cache
 ---> 75dc736cb568
Step 3/8 : VOLUME /report
 ---> Using cache
 ---> fc47f19eda0a
Step 4/8 : ENV FOCUS "(Networking).*(\[Conformance\])|\[Feature:NetworkPolicy\]"
 ---> Using cache
 ---> cdf6485c8b75
Step 5/8 : ADD kubeconfig /root/kubeconfig
 ---> Using cache
 ---> 0aa1f47f426b
Step 6/8 : ADD ./bin/e2e.test .
 ---> 6d3cea6190a1
Step 7/8 : ADD ./bin/kubectl /bin
 ---> 81e1412188d5
Step 8/8 : CMD ./e2e.test -kubeconfig=/root/kubeconfig --ginkgo.focus="$FOCUS" -report-dir=/report
 ---> Running in 15d30de72fb5
 ---> 1cf28e565007
Removing intermediate container 15d30de72fb5
Successfully built 1cf28e565007
Successfully tagged ozdanborne/k8s-e2e:latest
[dims@bigbox 09:40] ~/go/src/github.com/ozdanborne/k8s-e2e-containerized ⟩ kubectl delete pod e2e
Error from server (NotFound): pods "e2e" not found
[dims@bigbox 09:41] ~/go/src/github.com/ozdanborne/k8s-e2e-containerized ⟩ sudo kubectl run e2e --image=ozdanborne/k8s-e2e:latest --restart=Never --attach -- ./e2e.test --ginkgo.focus="Kubectl run job.*\[Conformance\]"
If you don't see a command prompt, try pressing enter.
Oct 16 13:41:13.261: INFO: 2 / 2 pods in namespace 'kube-system' are running and ready (0 seconds elapsed)
Oct 16 13:41:13.261: INFO: expected 2 pod replicas in namespace 'kube-system', 2 are Running and Ready.
Oct 16 13:41:13.263: INFO: Waiting for pods to enter Success, but no pods in "kube-system" match label map[name:e2e-image-puller]
Oct 16 13:41:13.263: INFO: Dumping network health container logs from all nodes...
Oct 16 13:41:13.265: INFO: Client version: v1.9.0-alpha.1.1132+3aba6451a75585
Oct 16 13:41:13.266: INFO: Server version: v1.9.0-alpha.1.1183+1ab5075c7c828a
SSSSSSSSSSSSSSSSSSSI1016 13:41:13.266692       1 e2e.go:63] The --provider flag is not set.  Treating as a conformance test.  Some tests may not be run.
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[sig-cli] Kubectl client [k8s.io] Kubectl run job
  should create a job from an image when restart is OnFailure [Conformance]
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:1371
[BeforeEach] [sig-cli] Kubectl client
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:133
STEP: Creating a kubernetes client
Oct 16 13:41:13.272: INFO: >>> kubeConfig: /tmp/kubeconfig-984649055
STEP: Building a namespace api object
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] [sig-cli] Kubectl client
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:286
[BeforeEach] [k8s.io] Kubectl run job
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:1348
[It] should create a job from an image when restart is OnFailure [Conformance]
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:1371
STEP: running the image gcr.io/google-containers/nginx-slim-amd64:0.20
Oct 16 13:41:13.374: INFO: Running '/bin/kubectl --kubeconfig=/tmp/kubeconfig-984649055 run e2e-test-nginx-job --restart=OnFailure --generator=job/v1 --image=gcr.io/google-containers/nginx-slim-amd64:0.20 --namespace=e2e-tests-kubectl-vrhrm'
Oct 16 13:41:16.051: INFO: stderr: ""
Oct 16 13:41:16.051: INFO: stdout: "job \"e2e-test-nginx-job\" created\n"
STEP: verifying the job e2e-test-nginx-job was created
[AfterEach] [k8s.io] Kubectl run job
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:1352
Oct 16 13:41:16.126: INFO: Running '/bin/kubectl --kubeconfig=/tmp/kubeconfig-984649055 delete jobs e2e-test-nginx-job --namespace=e2e-tests-kubectl-vrhrm'
Oct 16 13:41:18.466: INFO: stderr: ""
Oct 16 13:41:18.466: INFO: stdout: "job \"e2e-test-nginx-job\" deleted\n"
[AfterEach] [sig-cli] Kubectl client
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:134
Oct 16 13:41:18.466: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-tests-kubectl-vrhrm" for this suite.
Oct 16 13:41:24.566: INFO: namespace: e2e-tests-kubectl-vrhrm, resource: bindings, ignored listing per whitelist
Oct 16 13:41:24.586: INFO: namespace e2e-tests-kubectl-vrhrm deletion completed in 6.089442717s

• [SLOW TEST:11.314 seconds]
[sig-cli] Kubectl client
/home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/framework.go:22
  [k8s.io] Kubectl run job
  /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:634
    should create a job from an image when restart is OnFailure [Conformance]
    /home/dims/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl/kubectl.go:1371
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSOct 16 13:41:24.587: INFO: Running AfterSuite actions on all node
Oct 16 13:41:24.587: INFO: Running AfterSuite actions on node 1

Ran 1 of 709 Specs in 11.415 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 708 Skipped PASS
```

timothysc commented on test/e2e/framework/test_context.go:
```diff
@@ -296,12 +300,61 @@ func ViperizeFlags() {
 	if len(t.Host) == 0 && len(t.KubeConfig) == 0 {
-		t.Host = defaultHost
+		// Check if we can use the in-cluster config
+		if clusterConfig, err := restclient.InClusterConfig(); err == nil {
```
So there is a section that already exists in e2e/framework/util.go called LoadConfig which I think is a more appropriate location.  We're kind of hijacking the general case with this code, vs. only the shell commands.  

I'm debating if we want to do that or now.

timothysc reviewed (commented):
see other comment.

dims commented:
Recap some discussion on slack : 

```
timothysc [4:55 PM] @dims commented, and could chat here about it.
timothysc [4:55 PM] I don't know if we want to over-ride there, b/c that's what were doing.
timothysc [4:55 PM] Also if we are, then we should do it in LoadConfig.
dims [5:08 PM] i looked at that @U09R2P666, we seem to call framework.LoadConfig() in many locations, since AfterReadingAllFlags is called just once after the flags are parsed, i opted to do it here. Happy to move things around
timothysc [5:09 PM] You could logic gate LoadConfig only if not set then, and I'd be in favor of that.
dims [5:13 PM] @timothysc i could about 15 locations where LoadConfig is called, so add if conditions before all those calls? (http://dims.dynu.net/hound?q=LoadConfig%5C(&i=nope&files=&repos=kubernetes)
dims [5:13 PM] we do check if it is already set inside the method - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1940
timothysc [5:22 PM] blarg, let me look at this.. I'd like to clean it up if we can.
dims [5:32 PM] @timothysc :slightly_smiling_face: sounds good
dims [5:33 PM] @timothysc feel free to cp the code from my PR if you want to do a consolidated cleanup. i can help test later
```

dims commented:
/retest

timothysc commented:
k, I'm not going to have time.  Could you open another issue to cleanup kubeconfig handling in the e2e tests, and I'll green light this.

dims commented:
@timothysc see https://github.com/kubernetes/kubernetes/issues/54168

timothysc commented on test/e2e/framework/test_context.go:
```diff
@@ -296,12 +300,61 @@ func ViperizeFlags() {
 }
 
+func createKubeConfig(clientCfg *restclient.Config) *clientcmdapi.Config {
+	clusterNick := "cluster"
```
@sttts Are there any utilities for converting an in-cluster-config to an on-disk kubeconfig?  

It seems weird that there isn't a utility to do this.

dims replied:
unfortunately did not find any ... i picked this from a test case - https://github.com/kubernetes/kubernetes/blob/master/test/integration/examples/apiserver_test.go#L363

timothysc reviewed (approved):
/lgtm - lets get this rolling.

timothysc commented:
/cc @jpbetz

luxas commented:
Thanks a lot @dims for this :)

dims commented:
/retest
